packages/core/src/agent/getOutboundMessageContext.ts

```diff
@@ -0,0 +1,247 @@
+
+  agentContext.config.logger.debug(
+    `Looking for out-of-band record for connection-less exchange for message ${message.id}`
+  )
+  const outOfBandRepository = agentContext.dependencyManager.resolve(OutOfBandRepository)
+  const outOfBandService = agentContext.dependencyManager.resolve(OutOfBandService)
+
+  const outOfBandRecord = await outOfBandRepository.findSingleByQuery(agentContext, {
+    invitationRequestsThreadIds: [message.threadId],
+  })
+
+  // Check if valid
+  if (outOfBandRecord?.role === OutOfBandRole.Sender) {
+    // Extract ourService from the oob record if not on a previous message
+    if (!ourService) {
+      ourService = await outOfBandService.getResolvedServiceForOutOfBandServices(
+        agentContext,
+        outOfBandRecord.outOfBandInvitation.getServices()
+      )
+    }
+
+    if (!recipientService) {
+      throw new AriesFrameworkError(
+        `Could not find a service to send the message to. Please make sure the connection has a service or provide a service to send the message to.`
+      )
+    }
+
+    // We have created the oob record with a message, that message should be provided here as well
+    if (!lastSentMessage) {
+      throw new AriesFrameworkError('Must have lastSentMessage when out of band record has role Sender')
+    }
+  } else if (outOfBandRecord?.role === OutOfBandRole.Receiver) {
+    // Extract recipientService from the oob record if not on a previous message
+    if (!recipientService) {
+      recipientService = await outOfBandService.getResolvedServiceForOutOfBandServices(
+        agentContext,
+        outOfBandRecord.outOfBandInvitation.getServices()
+      )
+    }
+
+    if (lastSentMessage && !ourService) {
+      throw new AriesFrameworkError(
+        `Could not find a service to send the message to. Please make sure the connection has a service or provide a service to send the message to.`
+      )
+    }
+  }
+  // we either miss ourService (even though a message was sent) or we miss recipientService
+  // we check in separate if statements to provide a more specific error message
+  else {
+    if (lastSentMessage && !ourService) {
+      agentContext.config.logger.error(
+        `No out of band record associated and missing our service for connection-less exchange for message ${message.id}, while previous message has already been sent.`
+      )
+      throw new AriesFrameworkError(
+        `No out of band record associated and missing our service for connection-less exchange for message ${message.id}, while previous message has already been sent.`
+      )
+    }
+
+    if (!recipientService) {
+      agentContext.config.logger.error(
+        `No out of band record associated and missing recipient service for connection-less exchange for message ${message.id}.`
+      )
+      throw new AriesFrameworkError(
+        `No out of band record associated and missing recipient service for connection-less exchange for message ${message.id}.`
+      )
```

Comment:
Should the logger.error get an additional option to also throw when logged? I think I have seem this pattern before and it seems a bit redundant.

Reply:
Yeah maybe, but then you can't throw different error types. Agree that this is not ideal. 

Something like this?

```ts
agentContext.config.logger.error(`message`, undefined /* extra data to log */, { throwError: true })
```

Reply:
Yeah something like that. Maybe something worth considering if we ever rework the logger.